### PR TITLE
refactor(amf):   amf uecontext storage manager

### PR DESCRIPTION
**Pull request by ganeshg87**
test cases:
   - verified below cases using UERANSIM
          1. Registration,
          2. PDU session setup,
          3. PDU session release,
          4. Deregistration

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by ganeshg87** [comment]
UML Class Diagram

**Comment by ganeshg87** [comment]
1. Executed oai test, all UT cases look fine
 
image

2. Basic Registration, PDU Session Establishment, Release, Deregistration verified with UERANSIM
image
[mme_reg_12_Dec_21_R0.log](https://github.com/magma/magma/files/7664913/mme_reg_12_Dec_21_R0.log)

File: lte/gateway/c/core/oai/include/map.h [hunk 1]
```diff
@@ -179,6 +179,29 @@ struct map_s {
     }
   }
 
+  /***************************************************************************
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/include/map.h:183]: can we add a TODO on this file on line 150 to mark that `std::unordered_map` will be changed to `google::protobuf::Map` interface?

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp [hunk 2]
```diff
@@ -43,9 +43,9 @@ namespace magma5g {
 int amf_handle_security_complete_response(
     amf_ue_ngap_id_t ue_id, amf_nas_message_decode_status_t decode_status) {
   OAILOG_FUNC_IN(LOG_NAS_AMF);
-  ue_m5gmm_context_s* ue_mm_context = NULL;
-  amf_context_t* amf_ctx            = NULL;
-  int rc                            = RETURNok;
+  std::shared_ptr<ue_m5gmm_context_t> ue_mm_context;
+  amf_context_t* amf_ctx = NULL;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: let's use `nullptr` on cpp files

File: lte/gateway/c/core/oai/tasks/amf/Registration.cpp [hunk 3]
```diff
@@ -673,8 +674,8 @@ int amf_send_registration_accept(amf_context_t* amf_context) {
 
 static int registration_accept_t3550_handler(
     zloop_t* loop, int timer_id, void* arg) {
-  amf_context_t* amf_ctx                         = NULL;
-  ue_m5gmm_context_s* ue_amf_context             = NULL;
+  amf_context_t* amf_ctx = NULL;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/Registration.cpp]: let's use `nullptr`

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h [hunk 4]
```diff
@@ -336,6 +337,7 @@ typedef struct amf_context_s {
   ambr_t subscribed_ue_ambr;
   /* apn_config_profile: set by S6A UPDATE LOCATION ANSWER */
   apn_config_profile_t apn_config_profile;
+  std::weak_ptr<ue_m5gmm_context_s> ue_context;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]:
```suggestion
  std::weak_ptr<ue_m5gmm_context_s> ue_context_p;
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp [hunk 5]
```diff
@@ -418,7 +422,7 @@ int amf_idle_mode_procedure(amf_context_t* amf_ctx) {
  **                                                                        **
  **                                                                        **
  ***************************************************************************/
-void amf_free_ue_context(ue_m5gmm_context_s* ue_context_p) {
+void amf_free_ue_context(std::shared_ptr<ue_m5gmm_context_t> ue_context_p) {
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:211]: Update documentation of this function accordingly

File: lte/gateway/c/core/oai/tasks/amf/amf_security_mode_control.cpp [hunk 6]
```diff
@@ -122,11 +122,11 @@ nas_amf_smc_proc_t* nas5g_new_smc_procedure(amf_context_t* const amf_context) {
  ***************************************************************************/
 static int amf_security_request(nas_amf_smc_proc_t* const smc_proc) {
   OAILOG_FUNC_IN(LOG_NAS_AMF);
-  ue_m5gmm_context_s* ue_mm_context = NULL;
-  amf_context_t* amf_ctx            = NULL;
-  amf_sap_t amf_sap                 = {};
-  int rc                            = RETURNerror;
-  smc_proc->T3560.id                = NAS5G_TIMER_INACTIVE_ID;
+  std::shared_ptr<ue_m5gmm_context_t> ue_mm_context;
+  amf_context_t* amf_ctx = NULL;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_security_mode_control.cpp]:
```suggestion
  amf_context_t* amf_ctx = nullptr;
```

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp [hunk 7]
```diff
@@ -281,7 +282,7 @@ static int pdu_session_resource_release_t3592_handler(
   amf_ue_ngap_id = uepdu_id.ue_id;
   pdu_session_id = uepdu_id.pdu_id;
 
-  ue_m5gmm_context_s* ue_context =
+  std::shared_ptr<ue_m5gmm_context_t> ue_context =
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: you can use `auto` for these retrievals

File: lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+// id-AMF-UE-NGAP-ID <--> ue_m5gmm_context_t map
+bool AmfUeContextStorage::amf_insert_into_amfid_ue_context_map(
+    std::shared_ptr<ue_m5gmm_context_t> pContext) {
+  if (!pContext && (INVALID_AMF_UE_NGAP_ID == pContext->amf_ue_ngap_id)) {
+    return false;
+  }
+  amfid_ue_context_map.insert_or_update(pContext->amf_ue_ngap_id, pContext);
+  return true;
+}
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp:65]: fix formatting on this file

File: lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+AmfUeContextStorage::amf_get_from_gnbid_ue_context_map(
+    gnb_ue_ngap_id_t ue_gnb_id) {
+  std::shared_ptr<ue_m5gmm_context_t> ctx;
+  magma::map_rc_t status = gnbid_ue_context_map.get(ue_gnb_id, ctx);
+  return ctx;
+}
+
+// GUTI <---> ue_m5gmm_context_t map
+bool AmfUeContextStorage::amf_insert_into_guti_ue_context_map(
+    std::shared_ptr<ue_m5gmm_context_t> pContext) {
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp]: let's not mix variable naming convention here, let's stick to what we are using through OAI as of now, which is underscore separated variables

File: lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+ **                                                                        **
+ ***************************************************************************/
+std::shared_ptr<ue_m5gmm_context_t>
+AmfUeContextStorage::amf_create_new_ue_context(void) {
+  std::shared_ptr<ue_m5gmm_context_t> new_p =
+      std::make_shared<ue_m5gmm_context_t>();
+
+  if (!new_p) {
+    OAILOG_ERROR(LOG_AMF_APP, "Failed to allocate memory for UE context \n");
+    return NULL;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp]:
```suggestion
    return nullptr;
```

**Review by ardzoht — CHANGES_REQUESTED** [review]
Please also resolve current merge conflicts

File: lte/gateway/c/core/oai/include/map.h (hunk 1 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/include/map.h:183]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp (hunk 2 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 4 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 5 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:211]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_security_mode_control.cpp (hunk 6 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_security_mode_control.cpp]: addressed review comment resolved all occurrences in the same file

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp (hunk 7 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp (hunk 8 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp:65]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp (hunk 10 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp (hunk 9 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_ue_context_storage.cpp]: addressed review comment

**Pull request closed without merging**